core/arch/arm/plat-imx/sm_platform_handler.c

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2017 NXP
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <io.h>
+#include <kernel/cache_helpers.h>
+#include <kernel/tz_ssvce_pl310.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <sm/sm.h>
+#include <sm/optee_smc.h>
+
+#include "smc_sip.h"
+
+static bool imx_sip_handler (struct thread_smc_args *smc_args) {
+
+	uint16_t sip_func = OPTEE_SMC_FUNC_NUM(smc_args->a0);
+
+	switch (sip_func) {
+#ifdef CFG_PL310
+	case SIP_IMX_L2C310:
+		smc_args->a0 = OPTEE_SMC_RETURN_OK;
+		DMSG("SIP L2C310 val=%x addr=%x\n", smc_args->a1,
+		     smc_args->a2);
+		/*
+		 * Disable L2 is only needed in low power case,
+		 * the psci asm code will take care of disable.
+		 * So we just flush data here.
+		 */
+		if (smc_args->a1 == 0 && smc_args->a2 == PL310_CTRL) {
+			dcache_op_all(DCACHE_OP_CLEAN);
+			arm_cl2_cleanbyway(pl310_base());
+			break;
+		}
+		write32(smc_args->a1, pl310_base() + smc_args->a2);
```

Comment:
With this normal world can write to any address mapped by secure world. I suggest exposing functions instead of unlimited writing to a bunch of registers.

It could be worth investigating which DT bindings are used by normal world for PL310 control, or if there's something that can be done to improve the situation.

Reply:
Yes. I should check smc_args->a2 here. In linux kernel, there is only a l2c_write_sec, if it is implemented, kernel will use it to initialize PL310, if no, kernel will use write_relaxed to write registers.